### DISCUSSION TITLE
How do you debug your custom pipelines?

### DISCUSSION THREAD

**Discussion by almajo** [open-webui/pipelines#357] [Q&A]
I am setting up a WebUI, Pipelines, LiteLLM-Proxy, Database combination. The base setup is there and now I want to add my pipelines.
Locally for debugging I setup most using docker-compose and I was wondering how you guys use debuggers like VSCode to see what's going on in you custom code?

I was experimenting with a git clone of the repository but that didn't quite work because of the way the pipelines are inserted within the pipe directory.

How are you dealing with it? print debugging when mounting it directly into the console?

**Comment by SotirisKot** [comment]
Hello,

For debugging with VSCode I am doing the following:

1. Clone the pipelines repository
2. Add your pipelines under the pipelines folder
3. Add your breakpoints in your code
4. Use devcontainers ([VSCode and devcontainers](https://code.visualstudio.com/docs/devcontainers/containers))
5. My devcontainer.json looks like this:
```javascript
{
	"name": "Existing Dockerfile",
	"build": {
		// Sets the run context to one level up instead of the .devcontainer folder.
		"context": "..",
		// Update the 'dockerFile' property if you aren't using the standard 'Dockerfile' filename.
		"dockerfile": "../Dockerfile" // this is the Dockerfile in the pipelines repository.
	},
        "features": {
         	"ghcr.io/devcontainers-contrib/features/ruff:1": {}
         },
         "runArgs": ["--network", "host"]
}
```
6. My launch.json for debugging looks like this:
```javascript
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Debugger: FastAPI",
            "type": "debugpy",
            "request": "launch",
            "module": "uvicorn",
            "args": [
                "main:app",
                "--reload",
                "--port", "9099", 
                "--host", "0.0.0.0"
            ],
            "env": {
                //your env variables if you have. These env variables will go into valves.
            },
            "jinja": true
        }
    ]
}
```
7. Run debugging using the abovementioned launch.json as you would normally do. This will deploy the FastAPI of pipelines in the port you specified.
8. Log into your webui account, navigate to: Admin Panel --> Settings --> Connections and under 'Manage OpenAI API Connections', add the IP of the pipelines API that you just deployed.
9. When you sent a query from a chat session in webui, that triggers your pipe, the execution will stop at your breakpoints and you can debug normally.

I believe it is not the optimal way, but I can not think of anything else, hope it helps!

**Reply by almajo** [reply]
thanks for sharing your setup! I was thinking about going into the same direction but didn't quite manage to do so.
